src/dstore/pmix_esh.c

```diff
@@ -139,23 +210,367 @@ static inline const char *_unique_id(void)
+                    _esh_dir_del(name);
+                }
+            }
+            else {
+                unlink(name);
+            }
+        }
+        closedir(dir);
+    }
+    else {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        return PMIX_ERROR;
+    }
+
+    return rmdir(path);
+}
+
+static inline int _esh_tbls_init(void)
+{
+    pmix_status_t rc = PMIX_SUCCESS;
+    size_t idx;
+
+    if (NULL == (_ns_track_array = PMIX_NEW(pmix_value_array_t))) {
+        rc = PMIX_ERR_OUT_OF_RESOURCE;
+        PMIX_ERROR_LOG(rc);
+        goto err_exit;
+    }
+    if (PMIX_SUCCESS != (rc = pmix_value_array_init(_ns_track_array, sizeof(ns_track_elem_t)))){
+        PMIX_ERROR_LOG(rc);
+        goto err_exit;
+    }
+
+    if (NULL == (_session_array = PMIX_NEW(pmix_value_array_t))){
+            rc = PMIX_ERR_OUT_OF_RESOURCE;
+    }
+    if (PMIX_SUCCESS != (rc = pmix_value_array_init(_session_array, sizeof(session_t)))) {
+        goto err_exit;
+    }
+    if (PMIX_SUCCESS != (rc = pmix_value_array_set_size(_session_array, ESH_INIT_SESSION_TBL_SIZE))) {
+        goto err_exit;
+    }
+    for (idx = 0; idx < ESH_INIT_SESSION_TBL_SIZE; idx++) {
+        PMIX_CONSTRUCT(pmix_value_array_get_item(_session_array, idx), session_t);
+    }
+
+    if (NULL == (_ns_map_array = PMIX_NEW(pmix_value_array_t))) {
+        rc = PMIX_ERR_OUT_OF_RESOURCE;
+    }
+    if (PMIX_SUCCESS != (rc = pmix_value_array_init(_ns_map_array, sizeof(ns_map_t)))) {
+        goto err_exit;
+    }
+    if (PMIX_SUCCESS != (rc = pmix_value_array_set_size(_ns_map_array, ESH_INIT_NS_MAP_TBL_SIZE))) {
+        goto err_exit;
+    }
+        for (idx = 0; idx < ESH_INIT_SESSION_TBL_SIZE; idx++) {
+        PMIX_CONSTRUCT(pmix_value_array_get_item(_ns_map_array, idx), ns_map_t);
+    }
+
+
+err_exit:
+    return rc;
+}
+
+static inline void _esh_ns_map_cleanup(void)
+{
+    size_t idx;
+    size_t size;
+    ns_map_t *ns_map;
+
+    if (NULL == _ns_map_array) {
+        return;
+    }
+
+    size = pmix_value_array_get_size(_ns_map_array);
+    ns_map = PMIX_VALUE_ARRAY_GET_BASE(_ns_map_array, ns_map_t);
+
+    for (idx = 0; idx < size; idx++) {
+        if(ns_map[idx].in_use)
+            PMIX_DESTRUCT(&ns_map[idx]);
+    }
+
+    PMIX_RELEASE(_ns_map_array);
+    _ns_map_array = NULL;
+}
+
+static inline void _esh_sessions_cleanup(void)
+{
+    size_t idx;
+    size_t size;
+    session_t *s_tbl;
+
+    if (NULL == _session_array) {
+        return;
+    }
+
+    size = pmix_value_array_get_size(_session_array);
+    s_tbl = PMIX_VALUE_ARRAY_GET_BASE(_session_array, session_t);
+
+    for (idx = 0; idx < size; idx++) {
+        if(s_tbl[idx].in_use)
+            PMIX_DESTRUCT(&s_tbl[idx].super);
+    }
+
+    PMIX_RELEASE(_session_array);
+    _session_array = NULL;
+}
+
+static inline void _esh_ns_track_cleanup(void)
+{
+    if (NULL == _ns_track_array) {
+        return;
+    }
+
+    PMIX_DESTRUCT(_ns_track_array);
```

Comment:
This doesn't ensure that all elements are cleaned up properly.
In case of the pmix_list I guess that destructor for each element will be called, but here it is not a case.
Do we need to cleanup following fields here?
https://github.com/pmix/master/blob/master/src/dstore/pmix_esh.h#L74:L77